Cancel all observable meet exception

Old code:
```kotlin
import java.util.concurrent.TimeUnit
import rx.Observable

fun fetchFromNetwork(): Observable<String> {
    return Observable.timer(1, TimeUnit.HOURS).map { "Result from network" }
}

fun getResult(): Observable<Boolean> {
    return fetchFromNetwork().map { "network" in it }
}

val subscription = getResult().subscribe()
subscription.unsubscribe()
```

New code:
```kotlin
import kotlinx.coroutines.experimental.delay
import kotlinx.coroutines.experimental.launch
import kotlinx.coroutines.experimental.rx1.await
import kotlinx.coroutines.experimental.rx1.awaitFirst
import kotlinx.coroutines.experimental.rx1.rxObservable
import kotlinx.coroutines.experimental.rx1.rxSingle
import java.util.concurrent.TimeUnit

fun fetchFromNetwork() = rxObservable {
    delay(1, TimeUnit.HOURS)
    send("Result from network")
}

fun getResult() = rxObservable {
    val result = fetchFromNetwork().awaitFirst()
    send("network" in result)
}

val subscription = getResult().subscribe()
subscription.unsubscribe()
```

The old code will cancel job quietly, but the new code will throw an exception: 
```
Caused by: rx.exceptions.CompositeException$CompositeExceptionCausalChain: Chain of Causes for CompositeException In Order Received =>
	at rx.exceptions.CompositeException.getCause(CompositeException.java:129)
	at java.lang.Throwable.printEnclosedStackTrace(Throwable.java:707)
	at java.lang.Throwable.printStackTrace(Throwable.java:667)
	at java.lang.Throwable.printStackTrace(Throwable.java:643)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1061)
	at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1052)
	at kotlinx.coroutines.experimental.CoroutineExceptionHandlerImplKt.handleCoroutineExceptionImpl(CoroutineExceptionHandlerImpl.kt:30)
	at kotlinx.coroutines.experimental.CoroutineExceptionHandlerKt.handleCoroutineException(CoroutineExceptionHandler.kt:47)
	at kotlinx.coroutines.experimental.rx1.RxObservableCoroutine.doLockedSignalCompleted(RxObservable.kt:176)
	... 32 more
Caused by: java.util.NoSuchElementException: Sequence contains no elements
	at rx.internal.operators.OperatorSingle$ParentSubscriber.onCompleted(OperatorSingle.java:115)
	at rx.internal.operators.OperatorTake$1.onCompleted(OperatorTake.java:56)
	at kotlinx.coroutines.experimental.rx1.RxObservableCoroutine.doLockedSignalCompleted(RxObservable.kt:174)
	... 32 more
Caused by: java.lang.IllegalStateException: Unexpected update, state: CancelledContinuation[kotlinx.coroutines.experimental.JobCancellationException: Job was cancelled normally; job=RxObservableCoroutine{Cancelling}@73d7e407], update: CompletedExceptionally[java.util.NoSuchElementException: Sequence contains no elements]
	at kotlinx.coroutines.experimental.AbstractContinuation.resumeImpl(AbstractContinuation.kt:293)
	at kotlinx.coroutines.experimental.AbstractContinuation.resumeWithException(AbstractContinuation.kt:158)
	at kotlinx.coroutines.experimental.rx1.RxAwaitKt$awaitOne$2$1.onError(RxAwait.kt:135)
	at rx.observers.SafeSubscriber._onError(SafeSubscriber.java:153)
	at rx.observers.SafeSubscriber.onError(SafeSubscriber.java:115)
	... 35 more
```

Is my usage wrong?

--- a/common/kotlinx-coroutines-core-common/src/AbstractContinuation.kt
+++ b/common/kotlinx-coroutines-core-common/src/AbstractContinuation.kt
@@ -277,12 +277,16 @@ internal abstract class AbstractContinuation<in T>(
                 }
                 is CancelledContinuation -> {
                     /*
-                     * If continuation was cancelled, then all further updates (resumes or exceptions) must be
-                     * ignored, because cancellation is asynchronous and may race with resume/resumeWithException.
-                     * This race is normal.
+                     * If continuation was cancelled, then all further resumes must be
+                     * ignored, because cancellation is asynchronous and may race with resume.
+                     * Racy exception are reported so no exceptions are lost
                      *
                      * :todo: we should somehow remember the attempt to invoke resume and fail on the second attempt.
                      */
+                    if (proposedUpdate is CompletedExceptionally) {
+                        handleException(proposedUpdate.cause)
+                    }
+
                     return
                 }
                 else -> error("Already resumed, but proposed with update $proposedUpdate")

--- a/common/kotlinx-coroutines-core-common/src/AbstractContinuation.kt
+++ b/common/kotlinx-coroutines-core-common/src/AbstractContinuation.kt
@@ -272,15 +272,17 @@ internal abstract class AbstractContinuation<in T>(
                         }
                     }
                 }
-
                 is NotCompleted -> {
                     if (updateStateToFinal(state, proposedUpdate, resumeMode)) return
                 }
                 is CancelledContinuation -> {
-                    if (proposedUpdate is NotCompleted || proposedUpdate is CompletedExceptionally) {
-                        error("Unexpected update, state: $state, update: $proposedUpdate")
-                    }
-                    // Coroutine is dispatched normally (e.g.via `delay()`) after cancellation
+                    /*
+                     * If continuation was cancelled, then all further updates (resumes or exceptions) must be
+                     * ignored, because cancellation is asynchronous and may race with resume/resumeWithException.
+                     * This race is normal.
+                     *
+                     * :todo: we should somehow remember the attempt to invoke resume and fail on the second attempt.
+                     */
                     return
                 }
                 else -> error("Already resumed, but proposed with update $proposedUpdate")

--- a/common/kotlinx-coroutines-core-common/test/CancellableContinuationTest.kt
+++ b/common/kotlinx-coroutines-core-common/test/CancellableContinuationTest.kt
@@ -1,0 +1,120 @@
+/*
+ * Copyright 2016-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.experimental
+
+import kotlin.coroutines.experimental.*
+import kotlin.test.*
+
+class CancellableContinuationTest : TestBase() {
+    @Test
+    fun testResumeWithExceptionAndResumeWithException() = runTest {
+        var continuation: Continuation<Unit>? = null
+        val job = launch(coroutineContext) {
+            try {
+                expect(2)
+                suspendCancellableCoroutine<Unit> { c ->
+                    continuation = c
+                }
+            } catch (e: TestException) {
+                expect(3)
+            }
+        }
+        expect(1)
+        yield()
+        continuation!!.resumeWithException(TestException())
+        yield()
+        assertFailsWith<IllegalStateException> { continuation!!.resumeWithException(TestException()) }
+        job.join()
+        finish(4)
+    }
+
+    @Test
+    fun testResumeAndResumeWithException() = runTest {
+        var continuation: Continuation<Unit>? = null
+        val job = launch(coroutineContext) {
+            expect(2)
+            suspendCancellableCoroutine<Unit> { c ->
+                continuation = c
+            }
+            expect(3)
+        }
+        expect(1)
+        yield()
+        continuation!!.resume(Unit)
+        job.join()
+        assertFailsWith<IllegalStateException> { continuation!!.resumeWithException(TestException()) }
+        finish(4)
+    }
+
+    @Test
+    fun testResumeAndResume() = runTest {
+        var continuation: Continuation<Unit>? = null
+        val job = launch(coroutineContext) {
+            expect(2)
+            suspendCancellableCoroutine<Unit> { c ->
+                continuation = c
+            }
+            expect(3)
+        }
+        expect(1)
+        yield()
+        continuation!!.resume(Unit)
+        job.join()
+        assertFailsWith<IllegalStateException> { continuation!!.resume(Unit) }
+        finish(4)
+    }
+
+    /**
+     * Cancelling outer job may, in practise, race with attempt to resume continuation and resumes
+     * should be ignored. Here suspended coroutine is cancelled but then resumed with exception.
+     */
+    @Test
+    fun testCancelAndResumeWithException() = runTest {
+        var continuation: Continuation<Unit>? = null
+        val job = launch(coroutineContext) {
+            try {
+                expect(2)
+                suspendCancellableCoroutine<Unit> { c ->
+                    continuation = c
+                }
+            } catch (e: JobCancellationException) {
+                expect(3)
+            }
+        }
+        expect(1)
+        yield()
+        job.cancel() // Cancel job
+        yield()
+        continuation!!.resumeWithException(TestException()) // Should not fail
+        finish(4)
+    }
+
+    /**
+     * Cancelling outer job may, in practise, race with attempt to resume continuation and resumes
+     * should be ignored. Here suspended coroutine is cancelled but then resumed with exception.
+     */
+    @Test
+    fun testCancelAndResume() = runTest {
+        var continuation: Continuation<Unit>? = null
+        val job = launch(coroutineContext) {
+            try {
+                expect(2)
+                suspendCancellableCoroutine<Unit> { c ->
+                    continuation = c
+                }
+            } catch (e: JobCancellationException) {
+                expect(3)
+            }
+        }
+        expect(1)
+        yield()
+        job.cancel() // Cancel job
+        yield()
+        continuation!!.resume(Unit) // Should not fail
+        finish(4)
+    }
+
+    private class TestException : Exception()
+}

--- a/common/kotlinx-coroutines-core-common/test/CancellableContinuationTest.kt
+++ b/common/kotlinx-coroutines-core-common/test/CancellableContinuationTest.kt
@@ -1,6 +1,7 @@
 /*
  * Copyright 2016-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
  */
+@file:Suppress("NAMED_ARGUMENTS_NOT_ALLOWED") // KT-21913
 
 package kotlinx.coroutines.experimental
 
@@ -71,7 +72,7 @@ class CancellableContinuationTest : TestBase() {
      * should be ignored. Here suspended coroutine is cancelled but then resumed with exception.
      */
     @Test
-    fun testCancelAndResumeWithException() = runTest {
+    fun testCancelAndResumeWithException() = runTest(unhandled = listOf({e -> e is TestException})) {
         var continuation: Continuation<Unit>? = null
         val job = launch(coroutineContext) {
             try {

--- a/core/kotlinx-coroutines-core/test/CancellableContinuationExceptionHandlingTest.kt
+++ b/core/kotlinx-coroutines-core/test/CancellableContinuationExceptionHandlingTest.kt
@@ -215,53 +215,6 @@ class CancellableContinuationExceptionHandlingTest : TestBase() {
         }
     }
 
-    @Test
-    fun testMultipleCancellations() = runTest {
-        var continuation: Continuation<Unit>? = null
-
-        val job = launch(coroutineContext) {
-            try {
-                expect(2)
-                suspendCancellableCoroutine<Unit> { c ->
-                    continuation = c
-                }
-            } catch (e: IOException) {
-                expect(3)
-            }
-        }
-
-        expect(1)
-        yield()
-        continuation!!.resumeWithException(IOException())
-        yield()
-        assertFailsWith<IllegalStateException> { continuation!!.resumeWithException(ClosedChannelException()) }
-        try {
-            job.join()
-        } finally {
-            finish(4)
-        }
-    }
-
-    @Test
-    fun testResumeAndCancel() = runTest {
-        var continuation: Continuation<Unit>? = null
-
-        val job = launch(coroutineContext) {
-            expect(2)
-            suspendCancellableCoroutine<Unit> { c ->
-                continuation = c
-            }
-            expect(3)
-        }
-
-        expect(1)
-        yield()
-        continuation!!.resume(Unit)
-        job.join()
-        assertFailsWith<IllegalStateException> { continuation!!.resumeWithException(ClosedChannelException()) }
-        finish(4)
-    }
-
     private fun wrapperDispatcher(context: CoroutineContext): CoroutineContext {
         val dispatcher = context[ContinuationInterceptor] as CoroutineDispatcher
         return object : CoroutineDispatcher() {

--- a/reactive/kotlinx-coroutines-reactive/test/IntegrationTest.kt
+++ b/reactive/kotlinx-coroutines-reactive/test/IntegrationTest.kt
@@ -32,7 +32,7 @@ class IntegrationTest(
         @JvmStatic
         fun params(): Collection<Array<Any>> = Ctx.values().flatMap { ctx ->
             listOf(false, true).map { delay ->
-                arrayOf<Any>(ctx, delay)
+                arrayOf(ctx, delay)
             }
         }
     }
@@ -93,6 +93,34 @@ class IntegrationTest(
         val channel = pub.openSubscription()
         checkNumbers(n, channel.asPublisher(ctx(coroutineContext)))
         channel.cancel()
+    }
+
+    @Test
+    fun testCancelWithoutValue() = runTest {
+        val job = launch(coroutineContext, parent = Job(), start = CoroutineStart.UNDISPATCHED) {
+            publish<String>(coroutineContext) {
+                yield()
+                expectUnreached()
+            }.awaitFirst()
+        }
+
+        job.cancel()
+        job.join()
+    }
+
+    @Test
+    fun testEmptySingle() = runTest(unhandled = listOf({e -> e is NoSuchElementException})) {
+        expect(1)
+        val job = launch(coroutineContext, parent = Job(), start = CoroutineStart.UNDISPATCHED) {
+            publish<String>(coroutineContext) {
+                yield()
+                expect(2)
+                // Nothing to emit
+            }.awaitFirst()
+        }
+
+        job.join()
+        finish(3)
     }
 
     private suspend fun checkNumbers(n: Int, pub: Publisher<Int>) {

--- a/reactive/kotlinx-coroutines-rx1/src/RxAwait.kt
+++ b/reactive/kotlinx-coroutines-rx1/src/RxAwait.kt
@@ -120,7 +120,17 @@ private suspend fun <T> Observable<T>.awaitOne(): T = suspendCancellableCoroutin
         override fun onStart() { request(1) }
         override fun onNext(t: T) { cont.resume(t) }
         override fun onCompleted() { if (cont.isActive) cont.resumeWithException(IllegalStateException("Should have invoked onNext")) }
-        override fun onError(e: Throwable) { cont.resumeWithException(e) }
+        override fun onError(e: Throwable) {
+            /*
+             * Rx1 observable throws NoSuchElementException if cancellation happened before
+             * element emission. To mitigate this we try to atomically resume continuation with exception:
+             * if resume failed, then we know that continuation successfully cancelled itself
+             */
+            val token = cont.tryResumeWithException(e)
+            if (token != null) {
+                cont.completeResume(token)
+            }
+        }
     }))
 }
 

--- a/reactive/kotlinx-coroutines-rx1/test/IntegrationTest.kt
+++ b/reactive/kotlinx-coroutines-rx1/test/IntegrationTest.kt
@@ -32,7 +32,7 @@ class IntegrationTest(
         @JvmStatic
         fun params(): Collection<Array<Any>> = Ctx.values().flatMap { ctx ->
             listOf(false, true).map { delay ->
-                arrayOf<Any>(ctx, delay)
+                arrayOf(ctx, delay)
             }
         }
     }
@@ -115,6 +115,34 @@ class IntegrationTest(
         val channel = observable.openSubscription()
         checkNumbers(n, channel.asObservable(ctx(coroutineContext)))
         channel.cancel()
+    }
+
+    @Test
+    fun testCancelWithoutValue() = runTest {
+        val job = launch(coroutineContext, parent = Job(), start = CoroutineStart.UNDISPATCHED) {
+            rxObservable<String>(coroutineContext) {
+                yield()
+                expectUnreached()
+            }.awaitFirst()
+        }
+
+        job.cancel()
+        job.join()
+    }
+
+    @Test
+    fun testEmptySingle() = runTest(unhandled = listOf({e -> e is NoSuchElementException})) {
+        expect(1)
+        val job = launch(coroutineContext, parent = Job(), start = CoroutineStart.UNDISPATCHED) {
+            rxObservable<String>(coroutineContext) {
+                yield()
+                expect(2)
+                // Nothing to emit
+            }.awaitFirst()
+        }
+
+        job.join()
+        finish(3)
     }
 
     private suspend fun checkNumbers(n: Int, observable: Observable<Int>) {

--- a/reactive/kotlinx-coroutines-rx2/test/IntegrationTest.kt
+++ b/reactive/kotlinx-coroutines-rx2/test/IntegrationTest.kt
@@ -32,7 +32,7 @@ class IntegrationTest(
         @JvmStatic
         fun params(): Collection<Array<Any>> = Ctx.values().flatMap { ctx ->
             listOf(false, true).map { delay ->
-                arrayOf<Any>(ctx, delay)
+                arrayOf(ctx, delay)
             }
         }
     }
@@ -95,6 +95,34 @@ class IntegrationTest(
         val channel = observable.openSubscription()
         checkNumbers(n, channel.asObservable(ctx(coroutineContext)))
         channel.cancel()
+    }
+
+    @Test
+    fun testCancelWithoutValue() = runTest {
+        val job = launch(coroutineContext, parent = Job(), start = CoroutineStart.UNDISPATCHED) {
+            rxObservable<String>(coroutineContext) {
+                yield()
+                expectUnreached()
+            }.awaitFirst()
+        }
+
+        job.cancel()
+        job.join()
+    }
+
+    @Test
+    fun testEmptySingle() = runTest(unhandled = listOf({e -> e is NoSuchElementException})) {
+        expect(1)
+        val job = launch(coroutineContext, parent = Job(), start = CoroutineStart.UNDISPATCHED) {
+            rxObservable<String>(coroutineContext) {
+                yield()
+                expect(2)
+                // Nothing to emit
+            }.awaitFirst()
+        }
+
+        job.join()
+        finish(3)
     }
 
     private suspend fun checkNumbers(n: Int, observable: Observable<Int>) {

